fuse/ipns/ipns_unix.go

```diff
@@ -309,279 +271,130 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
-	}
 	return nil
 }
 
-// Signal that a node in this tree was changed so the root can republish
-func (n *Node) wasChanged() {
-	if IpnsReadonly {
-		return
-	}
-	root := n.nsRoot
-	if root == nil {
-		root = n
-	}
-
-	root.repub.Publish <- struct{}{}
+func (fi *File) Flush(ctx context.Context, req *fuse.FlushRequest) error {
+	return fi.fi.Close()
 }
 
-func (n *Node) republishRoot() error {
-
-	// We should already be the root, this is just a sanity check
-	var root *Node
-	if n.nsRoot != nil {
-		root = n.nsRoot
-	} else {
-		root = n
-	}
+func (fi *File) Fsync(ctx context.Context, req *fuse.FsyncRequest) error {
+	return fi.fi.Flush()
+}
 
-	// Add any nodes that may be new to the DAG service
-	err := n.Ipfs.DAG.AddRecursive(root.Nd)
+func (fi *File) Forget() {
+	err := fi.fi.Flush()
 	if err != nil {
-		log.Criticalf("ipns: Dag Add Error: %s", err)
-		return err
+		log.Debug("Forget file error: ", err)
 	}
+}
 
-	ndkey, err := root.Nd.Key()
+func (dir *Directory) Mkdir(ctx context.Context, req *fuse.MkdirRequest) (fs.Node, error) {
+	child, err := dir.dir.Mkdir(req.Name)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	err = n.Ipfs.Namesys.Publish(n.Ipfs.Context(), root.key, ndkey)
-	if err != nil {
-		return err
-	}
-	return nil
+	return &Directory{dir: child}, nil
 }
 
-func (n *Node) Fsync(ctx context.Context, req *fuse.FsyncRequest) error {
-	return nil
-}
-
-func (n *Node) Mkdir(ctx context.Context, req *fuse.MkdirRequest) (fs.Node, error) {
-	if IpnsReadonly {
-		return nil, fuse.EPERM
-	}
-	dagnd := &mdag.Node{Data: ft.FolderPBData()}
-	nnode := n.Nd.Copy()
-	nnode.AddNodeLink(req.Name, dagnd)
-
-	child := &Node{
-		Ipfs: n.Ipfs,
-		Nd:   dagnd,
-		name: req.Name,
-	}
-
-	if n.nsRoot == nil {
-		child.nsRoot = n
-	} else {
-		child.nsRoot = n.nsRoot
-	}
-
-	if n.parent != nil {
-		err := n.parent.update(n.name, nnode)
+func (fi *File) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fs.Handle, error) {
+	if req.Flags&fuse.OpenTruncate != 0 {
+		log.Warning("Need to truncate file!")
+		err := fi.fi.Truncate(0)
 		if err != nil {
-			log.Criticalf("Error updating node: %s", err)
 			return nil, err
 		}
-	}
-	n.Nd = nnode
-
-	n.wasChanged()
-
-	return child, nil
-}
-
-func (n *Node) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fs.Handle, error) {
-	//log.Debug("[%s] Received open request! flags = %s", n.name, req.Flags.String())
-	//TODO: check open flags and truncate if necessary
-	if req.Flags&fuse.OpenTruncate != 0 {
-		log.Warning("Need to truncate file!")
-		n.cached = nil
-		n.Nd = &mdag.Node{Data: ft.FilePBData(nil, 0)}
 	} else if req.Flags&fuse.OpenAppend != 0 {
 		log.Warning("Need to append to file!")
 	}
-	return n, nil
+	return fi, nil
 }
 
-func (n *Node) Mknod(ctx context.Context, req *fuse.MknodRequest) (fs.Node, error) {
-	return nil, nil
+func (fi *File) Release(ctx context.Context, req *fuse.ReleaseRequest) error {
+	return fi.fi.Close()
```

Comment:
is there anything to clear? what happens if fuse sends us operations after `Release`? will we fail gracefully or panic?